README.md
tableau_server Cookbook
================
[![Build Status](https://travis-ci.org/reevoo/tableau_server-chef.svg?branch=master)](https://travis-ci.org/reevoo/tableau_server-chef)

Installs Tableau Server

Usage
-----
#### `tableau::server`
Installs Tableau Server

* Create the tableau server encrypted databag for the chef environment you are targeting
* Configure the product key
e.g.

```json
{
  "product_key": "NOT-A-REAL-PRODUCT-KEY-X99"
}
```
* Add `tableau::server` to your node's run list

#### `tableau_server::redshift`
Installs the redshift driver and sets up an ODBC dsn.

* Configure the `['tableau_server']['redshift']['name']`, `['tableau_server']['redshift']['database']` and `['tableau_server']['redshift']['host']` attributes.
* Add the username and password to the tableau_server databag
e.g.

```json
{
  "redshift": {
    "username": "susan",
    "password": "youWiLLNEverGuess"
  }
}
```
* Add `tableau::redshift` to your node's run list

#### `tableau_server::backup`
Sets up a (nightly) task to backup Tableau Server to a s3 bucket

* Configure the `['tableau_server']['backup']['aws']['s3_bucket']` and `['tableau_server']['backup']['aws']['region']` attributes.
* Add the aws keys to the tableau_server databag
e.g.

```json
{
  "backup": {
    "aws_access_key": "NOTAREALKEY",
    "aws_secret_access_key": "NOTEVENREMOTLYSECRET"
  }
}
```
* Add `tableau::backup` to your node's run list


Requirements
------------
#### cookbooks
- [`windows`](https://supermarket.chef.io/cookbooks/windows)
- [`windows_firewall`] (https://supermarket.chef.io/cookbooks/windows_firewall)

Attributes
----------
#### `tableau_server::server`

| Key | Type | Description | Default |
|-----|------|-------------|---------|
| `['tableau_server']['version']` | string | The major and minor version of tableau server to install | `'9.2'` |
| `['tableau_server']['build']` | string | The build number of tableau server to install | `'9200.16.0204.1543'` |
| `['tableau_server']['source']` | string | the uri or path to install tableau from [see](http://www.tableau.com/support/esdalt) | `'https://downloads.tableau.com/esdalt/9.2.4/TableauServer-64bit-9-2-4.exe'` |
| `['tableau_server']['checksum']` | string | sha256 checksum for the installer | `'e21048d8ec59327ae0d05c313ad68efab59826d87c5549122001903ffc7d2583'` |
##### encrypted data bag `tableau_server <chef_environment>`
| Key | Type | Description |
|-----|------|-------------|
| `product_key` | string | Your tableau server product key |

#### `tableau_server::redshift`

| Key | Type | Description | Default |
|-----|------|-------------|---------|
| `['tableau_server']['redshift']['name']` | string | The name for the redshift ODBC DSN | `nil` |
| `['tableau_server']['redshift']['database']` | string | the name of the database | `nil` |
| `['tableau_server']['redshift']['host']` | string | the hostname of your redshift cluster | `nil` |
| `['tableau_server']['redshift']['port']` | string | the redshift port number | `'5439'` |
| `['tableau_server']['redshift']['proxy']['host']` | string | proxy hostname (if used) | `''` |
| `['tableau_server']['redshift']['proxy']['port']` | string | proxy port number (if used) | `''` |
| `['tableau_server']['redshift']['ssl_mode']` | string | ssl mode | `'require'` |
| `['tableau_server']['redshift']['bools_as_char']` | string | | `'1'` |
| `['tableau_server']['redshift']['fetch']` | string | | `'100'` |
| `['tableau_server']['redshift']['max_bytea']` | string | | `'255'` |
| `['tableau_server']['redshift']['max_long_varchar']` | string | | `'8190'` |
| `['tableau_server']['redshift']['max_varchar']` | string | | `'255'` |
| `['tableau_server']['redshift']['single_row_mode']` | string | | `'0'` |
| `['tableau_server']['redshift']['text_as_long_varchar']` | string | | `'1'` |
| `['tableau_server']['redshift']['use_declare_fetch']` | string | | `'0'` |
| `['tableau_server']['redshift']['use_multiple_statements']` | string | | `'0'` |
| `['tableau_server']['redshift']['use_unicode']` | string | | `'1'` |
##### encrypted data bag `tableau_server <chef_environment>`
| Key | Type | Description |
|-----|------|-------------|
| `['redshift']['username']` | string | the username for your redshift cluster |
| `['redshift']['password']` | string | the password for your redshift cluster |


#### `tableau_server::backup`

| Key | Type | Description | Default |
|-----|------|-------------|---------|
| `['tableau_server']['backup']['frequency']` | symbol | How often to backup tableau one of `:minute, :hourly, :daily, :weekly, :monthly, :once, :on_logon, :onstart, :on_idle` | `:daily` |
| `['tableau_server']['backup']['time']` | string | Specifies the start time to run the backup (HH:mm) | `'23:00'` |
| `['tableau_server']['backup']['aws']['source']` | string | The source uri or path to install the aws cli tools from  | `"https://s3.amazonaws.com/aws-cli/AWSCLI64.msi"` |
| `['tableau_server']['backup']['aws']['checksum']` | string | Checksum for the cli tools installer | ` "0385f9041a33d0ff9f7b954e12c0bec8f67bef1e30f990394f5fc784162fafdd"` |
| `['tableau_server']['backup']['aws']['s3_bucket']` | string | the name of the s3 bucket to backup to | `nil` |
| `['tableau_server']['backup']['aws']['region']` | string | the aws region of the backup bucket | `nil` |
##### encrypted data bag `tableau_server <chef_environment>`
| Key | Type | Description |
|-----|------|-------------|
| `['backup']['aws_access_key']` | string | The aws_access_key_id with access to the backup bucket |
| `['backup']['aws_secret_access_key']` | string | The aws_secret_access_key with access to the bucket |


Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch
3. Write your change(s)
4. Check everything works
5. Submit a Pull Request using Github

Bugs / Issues / Feature Requests
--------------------------------
Open an issue here on github

Todo / Roadmap
--------------
* Test Kitchen / CI (If it's possible on windows)
* Configure All The Things

Contributors
------------
Ed Robinson <ed@reevoo.com>


CHANGELOG.md
tableau_server CHANGELOG
=================

This file is used to list changes made in each version of the tableau_server cookbook.

0.1.0
-----
- [Ed Robinson] - Initial release of tableau_server cookbook


recipes/backup.rb
if node.tableau_server["backup"]["aws"]["s3_bucket"] && node.tableau_server["backup"]["aws"]["region"]
  directory 'C:\Program Files\Tableau\Backups'

  package "AWS Tools for Windows" do
    source node.tableau_server["backup"]["aws"]["source_uri"]
    checksum node.tableau_server["backup"]["aws"]["checksum"]
    action :install
  end

  backup = Chef::EncryptedDataBagItem.load("tableau_server", node.chef_environment)["backup"]

  directory File.join(Dir.home, ".aws")

  file File.join(Dir.home, '.aws\credentials') do
    content %([default]
aws_access_key_id = #{backup["aws_access_key"]}
aws_secret_access_key = #{backup["aws_secret_access_key"]})
  end

  file File.join(Dir.home, '.aws\config') do
    content "[default]\nregion = #{node.tableau_server["backup"]["aws"]["region"]}"
  end

  template 'C:Program Files\Tableau\backup.bat' do
    source "backup.bat.erb"
    variables(
      aws_access_key_id: backup["aws_access_key"],
      aws_secret_access_key: backup["aws_secret_access_key"],
    )
  end

  windows_task "Tableau Backup" do
    command "backup.bat"
    cwd 'C:\Program Files\Tableau'
    frequency node.tableau_server["backup"]["frequency"]
    start_time node.tableau_server["backup"]["time"]
    action [:create, :enable]
  end
end


recipes/server.rb
package "Tableau Server #{node.tableau_server["version"]} (#{node.tableau_server["build"]})" do
  source node.tableau_server["source"]
  version node.tableau_server["version"]
  installer_type :custom
  action :install
  checksum node.tableau_server["checksum"]
  options "/verysilent"
end

databag = Chef::EncryptedDataBagItem.load("tableau_server", node.chef_environment)

# It is a pain that we have to run this every chef run but I can't find a
# way to check if we have already registered.
execute "Register Tableau Server" do
  command "tabinstallck.exe -activate #{databag["product_key"]}"
  cwd 'C:\Program Files\Tableau\Tableau Server\\' + node.tableau_server["version"] + '\bin'
end

windows_firewall_rule "Tableau Server" do
  protocol "TCP"
  localport "80"
  firewall_action :allow
end

windows_firewall_rule "Tableau Server (HTTPS)" do
  protocol "TCP"
  localport "443"
  firewall_action :allow
end


recipes/redshift.rb
redshift = node.tableau_server["redshift"]

if redshift["name"] && redshift["host"] && redshift["database"]
  package redshift["driver"]["name"] do
    source redshift["driver"]["source_uri"]
    checksum redshift["driver"]["checksum"]
  end

  databag = Chef::EncryptedDataBagItem.load("tableau_server", node.chef_environment)["redshift"]

  registry_key "HKEY_LOCAL_MACHINE\\SOFTWARE\\ODBC\\ODBC.INI\\#{redshift["name"]}" do
    values [
      { type: :string, name: "Driver",                data: redshift["driver"]["name"] },
      { type: :string, name: "Database",              data: redshift["database"] },
      { type: :string, name: "UID",                   data: databag["username"] },
      { type: :string, name: "PWD",                   data: databag["password"] },
      { type: :string, name: "Server",                data: redshift["host"] },
      { type: :string, name: "Port",                  data: redshift["port"] },
      { type: :string, name: "ProxyHost",             data: redshift["proxy"]["host"] },
      { type: :string, name: "ProxyPort",             data: redshift["proxy"]["port"] },
      { type: :string, name: "SSLMode",               data: redshift["ssl_mode"] },
      { type: :string, name: "BoolsAsChar",           data: redshift["bools_as_char"] },
      { type: :string, name: "Fetch",                 data: redshift["fetch"] },
      { type: :string, name: "MaxBytea",              data: redshift["max_bytea"] },
      { type: :string, name: "MaxLongVarchar",        data: redshift["max_long_varchar"] },
      { type: :string, name: "MaxVarchar",            data: redshift["max_varchar"] },
      { type: :string, name: "SingleRowMode",         data: redshift["single_row_mode"] },
      { type: :string, name: "TextAsLongVarchar",     data: redshift["text_as_long_varchar"] },
      { type: :string, name: "UseDeclareFetch",       data: redshift["use_declare_fetch"] },
      { type: :string, name: "UseMultipleStatements", data: redshift["use_multiple_statements"] },
      { type: :string, name: "UseUnicode",            data: redshift["use_unicode"] },
    ]
  end
end


attributes/default.rb
default["tableau_server"]["version"] = "9.2"
default["tableau_server"]["build"] = "9200.16.0204.1543"
default["tableau_server"]["source"] = "https://downloads.tableau.com/esdalt/9.2.4/TableauServer-64bit-9-2-4.exe"
default["tableau_server"]["checksum"] = "e21048d8ec59327ae0d05c313ad68efab59826d87c5549122001903ffc7d2583"

default["tableau_server"]["redshift"]["driver"]["name"]       = "Amazon Redshift (x64)"
default["tableau_server"]["redshift"]["driver"]["source_uri"] = "https://s3.amazonaws.com" \
                         "/redshift-downloads/drivers/AmazonRedshiftODBC64-1.2.6.1006.msi"
default["tableau_server"]["redshift"]["driver"]["checksum"]   = "23530d6215140f0f8b06c128" \
                                                "68c5ce271485a95cbfbdfc1e4479a2b4f8e5532c"

default["tableau_server"]["redshift"]["name"] = nil
default["tableau_server"]["redshift"]["database"] = nil
default["tableau_server"]["redshift"]["host"] = nil
default["tableau_server"]["redshift"]["port"] = "5439"
default["tableau_server"]["redshift"]["proxy"]["host"] = ""
default["tableau_server"]["redshift"]["proxy"]["port"] = ""
default["tableau_server"]["redshift"]["ssl_mode"]                = "require"
default["tableau_server"]["redshift"]["bools_as_char"]           = "1"
default["tableau_server"]["redshift"]["fetch"]                   = "100"
default["tableau_server"]["redshift"]["max_bytea"]               = "255"
default["tableau_server"]["redshift"]["max_long_varchar"]        = "8190"
default["tableau_server"]["redshift"]["max_varchar"]             = "255"
default["tableau_server"]["redshift"]["single_row_mode"]         = "0"
default["tableau_server"]["redshift"]["text_as_long_varchar"]    = "1"
default["tableau_server"]["redshift"]["use_declare_fetch"]       = "0"
default["tableau_server"]["redshift"]["use_multiple_statements"] = "0"
default["tableau_server"]["redshift"]["use_unicode"]             = "1"

default["tableau_server"]["backup"]["frequency"] = :daily
default["tableau_server"]["backup"]["time"] = "23:00"
default["tableau_server"]["backup"]["aws"]["source_uri"] = "https://s3.amazonaws.com/aws-cli/AWSCLI64.msi"
default["tableau_server"]["backup"]["aws"]["checksum"] = "0385f9041a33d0ff9f7b954e12c0bec" \
                                                       "8f67bef1e30f990394f5fc784162fafdd"
default["tableau_server"]["backup"]["aws"]["s3_bucket"] = nil
default["tableau_server"]["backup"]["aws"]["region"] = nil
